File: clients/cpp/MyStrategy.cpp
#include "MyStrategy.hpp"

MyStrategy::MyStrategy() {}

double distanceSqr(Vec2Double a, Vec2Double b) {
  return (a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y);
}

UnitAction MyStrategy::getAction(const Unit &unit, const Game &game,
                                 Debug &debug) {
  const Unit *nearestEnemy = nullptr;
  for (const Unit &other : game.units) {
    if (other.playerId != unit.playerId) {
      if (nearestEnemy == nullptr ||
          distanceSqr(unit.position, other.position) <
              distanceSqr(unit.position, nearestEnemy->position)) {
        nearestEnemy = &other;
      }
    }
  }
  const LootBox *nearestWeapon = nullptr;
  for (const LootBox &lootBox : game.lootBoxes) {
    if (std::dynamic_pointer_cast<Item::Weapon>(lootBox.item)) {
      if (nearestWeapon == nullptr ||
          distanceSqr(unit.position, lootBox.position) <
              distanceSqr(unit.position, nearestWeapon->position)) {
        nearestWeapon = &lootBox;
      }
    }
  }
  Vec2Double targetPos = unit.position;
  if (unit.weapon == nullptr && nearestWeapon != nullptr) {
    targetPos = nearestWeapon->position;
  } else if (nearestEnemy != nullptr) {
    targetPos = nearestEnemy->position;
  }
  debug.draw(CustomData::Log(
      std::string("Target pos: ") + targetPos.toString()));
  Vec2Double aim = Vec2Double(0, 0);
  if (nearestEnemy != nullptr) {
    aim = Vec2Double(nearestEnemy->position.x - unit.position.x,
                     nearestEnemy->position.y - unit.position.y);
  }
  bool jump = targetPos.y > unit.position.y;
  if (targetPos.x > unit.position.x &&
      game.level.tiles[size_t(unit.position.x + 1)][size_t(unit.position.y)] ==
          Tile::WALL) {
    jump = true;
  }
  if (targetPos.x < unit.position.x &&
      game.level.tiles[size_t(unit.position.x - 1)][size_t(unit.position.y)] ==
          Tile::WALL) {
    jump = true;
  }
  UnitAction action;
  action.velocity = targetPos.x - unit.position.x;
  action.jump = jump;
  action.jumpDown = !action.jump;
  action.aim = aim;
  action.shoot = true;
  action.swapWeapon = false;
  action.plantMine = false;
  return action;
}

File: clients/rust/src/my_strategy.rs
pub struct MyStrategy {}

impl MyStrategy {
    pub fn new() -> Self {
        Self {}
    }
    pub fn get_action(
        &mut self,
        unit: &model::Unit,
        game: &model::Game,
        debug: &mut crate::Debug,
    ) -> model::UnitAction {
        fn distance_sqr(a: &model::Vec2F64, b: &model::Vec2F64) -> f64 {
            (a.x - b.x).powi(2) + (a.y - b.y).powi(2)
        }
        let nearest_enemy = game
            .units
            .iter()
            .filter(|other| other.player_id != unit.player_id)
            .min_by(|a, b| {
                std::cmp::PartialOrd::partial_cmp(
                    &distance_sqr(&a.position, &unit.position),
                    &distance_sqr(&b.position, &unit.position),
                )
                .unwrap()
            });
        let nearest_weapon = game
            .loot_boxes
            .iter()
            .filter(|loot| {
                if let model::Item::Weapon { .. } = loot.item {
                    true
                } else {
                    false
                }
            })
            .min_by(|a, b| {
                std::cmp::PartialOrd::partial_cmp(
                    &distance_sqr(&a.position, &unit.position),
                    &distance_sqr(&b.position, &unit.position),
                )
                .unwrap()
            });
        let mut target_pos = unit.position.clone();
        if let (&None, Some(weapon)) = (&unit.weapon, nearest_weapon) {
            target_pos = weapon.position.clone();
        } else if let Some(enemy) = nearest_enemy {
            target_pos = enemy.position.clone();
        }
        debug.draw(model::CustomData::Log {
            text: format!("Target pos: {:?}", target_pos),
        });
        let mut aim = model::Vec2F64 { x: 0.0, y: 0.0 };
        if let Some(enemy) = nearest_enemy {
            aim = model::Vec2F64 {
                x: enemy.position.x - unit.position.x,
                y: enemy.position.y - unit.position.y,
            };
        }
        let mut jump = target_pos.y > unit.position.y;
        if target_pos.x > unit.position.x
            && game.level.tiles[(unit.position.x + 1.0) as usize][(unit.position.y) as usize]
                == model::Tile::Wall
        {
            jump = true
        }
        if target_pos.x < unit.position.x
            && game.level.tiles[(unit.position.x - 1.0) as usize][(unit.position.y) as usize]
                == model::Tile::Wall
        {
            jump = true
        }
        model::UnitAction {
            velocity: target_pos.x - unit.position.x,
            jump,
            jump_down: target_pos.y < unit.position.y,
            aim,
            shoot: true,
            swap_weapon: false,
            plant_mine: false,
        }
    }
}
